features/FEATURE_BLE/ble/SecurityManager.h

```diff
@@ -74,22 +88,201 @@ class SecurityManager {
-    typedef void (*PasskeyDisplayCallback_t)(Gap::Handle_t handle, const Passkey_t passkey);
+    typedef uint8_t Passkey_t[PASSKEY_LEN]; /**< 6-digit passkey in ASCII ('0'-'9' digits only). */
 
     typedef FunctionPointerWithContext<const SecurityManager *> SecurityManagerShutdownCallback_t;
     typedef CallChainOfFunctionPointersWithContext<const SecurityManager *> SecurityManagerShutdownCallbackChain_t;
 
+    /* legacy callbacks, please use SecurityManagerEventHandler instead */
+    typedef void (*HandleSpecificEvent_t)(ble::connection_handle_t connectionHandle);
+    typedef void (*SecuritySetupInitiatedCallback_t)(ble::connection_handle_t, bool allowBonding, bool requireMITM, SecurityIOCapabilities_t iocaps);
+    typedef void (*SecuritySetupCompletedCallback_t)(ble::connection_handle_t, SecurityCompletionStatus_t status);
+    typedef void (*LinkSecuredCallback_t)(ble::connection_handle_t connectionHandle, SecurityMode_t securityMode);
+    typedef void (*PasskeyDisplayCallback_t)(ble::connection_handle_t connectionHandle, const Passkey_t passkey);
+
+    /** The stack will use these functions to signal events to the application,
+     *  subclass to override handlers. Use SecurityManager::setSecurityManagerEventHandler
+     *  to set the interface implementation to be used. */
+    class SecurityManagerEventHandler {
+    public:
+        SecurityManagerEventHandler() {};
+        virtual ~SecurityManagerEventHandler() {};
+
+        ////////////////////////////////////////////////////////////////////////////
+        // Pairing
+        //
+
+        /**
+         * Request application to accept or reject pairing. Application should respond by
+         * calling the appropriate function: acceptPairingRequest or cancelPairingRequest
+         *
+         * @param[in] connectionHandle connection connectionHandle
+         */
+        virtual void pairingRequest(ble::connection_handle_t connectionHandle) {
+            (void)connectionHandle;
+        }
+
+        /**
+         * Indicate to the application that pairing has completed.
+         *
+         * @param[in] connectionHandle connection connectionHandle
+         * @param[in] result result of the pairing indicating success or reason for failure
+         */
+        virtual void pairingResult(ble::connection_handle_t connectionHandle, SecurityCompletionStatus_t result) {
+            (void)connectionHandle;
+            (void)result;
+        }
+
+        ////////////////////////////////////////////////////////////////////////////
+        // Security
+        //
+
+        /**
+         * Deliver the requested whitelist to the application.
+         *
+         * @param[in] whitelist whitelist created based on the bonding table,
+         *                      caller transfers the memory ownership
+         */
+        virtual void whitelistFromBondTable(Gap::Whitelist_t* whitelist) {
+            if (whitelist) {
```

Comment:
This block is dangerous. Nothings indicate the whitelist has been dynamically allocated.